### PR TITLE
MAINT: switch backend to github actions rather than EC2

### LINKED ISSUE #359: INFR: Move from EC2 backend to GitHub Actions
@mmcky needs to update this repo to build on `GitHub actions` instead of `EC2`

This will improve compatibility with fork contributions and reduce costs.

### PR DESCRIPTION
**Pull request by mmcky**
This PR migrates from `EC2` powered backend to `GitHub Actions` as `jax` and `gpu` are no longer required. 

Fixes #359 

Review the following lectures for `jax` dependencies

- [ ] ar1_bayes
- [ ] back_prop
- [ ] bayes_nonconj
- [ ] mix_model

**Checks:**

- [ ] remove the cache and check a full build executes properly before merge

### CODE REVIEW

**Comment by mmcky** [comment]
@HumphreyYang it doesn't look like this series has been completely sanitised from `jax`

- [ ] lectures/ar1_bayes.md:!pip install arviz pymc numpyro jax
- [ ] lectures/ar1_bayes.md:import jax.numpy as jnp
- [ ] lectures/ar1_bayes.md:from jax import random
- [ ] lectures/back_prop.md:!pip install --upgrade jax jaxlib
- [ ] lectures/back_prop.md:import jax.numpy as jnp
- [ ] lectures/back_prop.md:from jax import grad, jit, jacfwd, vmap
- [ ] lectures/back_prop.md:from jax import random
- [ ] lectures/back_prop.md:import jax
- [ ] lectures/back_prop.md:    h = jax.nn.sigmoid
- [ ] lectures/back_prop.md:D = jax.scipy.linalg.block_diag(*[row.reshape((1, 2)) for row in jnp.block([[δ * xs[:-1]], [δ]]).T])
- [ ] lectures/back_prop.md:dxs_la = jax.scipy.linalg.solve_triangular(jnp.eye(N) - L, D, lower=True)
- [ ] lectures/back_prop.md:    D = jax.scipy.linalg.block_diag(*[row.reshape((1, 2)) for row in jnp.block([[δ * xs[:-1]], [δ]]).T])
- [ ] lectures/back_prop.md:    dxs_la = jax.scipy.linalg.solve_triangular(jnp.eye(N) - L, D, lower=True)
- [ ] lectures/back_prop.md:from jax.lib import xla_bridge
- [ ] lectures/bayes_nonconj.md:!pip install numpyro pyro-ppl torch jax
- [ ] lectures/bayes_nonconj.md:# jax
- [ ] lectures/bayes_nonconj.md:import jax.numpy as jnp
- [ ] lectures/bayes_nonconj.md:from jax import lax, random
- [ ] lectures/bayes_nonconj.md:We use both the packages `pyro` and `numpyro` with assistance from  `jax` to approximate a  posterior distribution
- [ ] lectures/bayes_nonconj.md:        # jax requires explicit PRNG state to be passed
- [ ] lectures/geom_series.md:init_printing(use_latex='mathjax')
- [ ] lectures/mix_model.md:import jax.numpy as jnp
- [ ] lectures/mix_model.md:from jax import random

We will need to do that before this can be merged. Can I get your help in coordinating this?

**Comment by HumphreyYang** [comment]
> @HumphreyYang it doesn't look like this series has been completely sanitised from `jax`
> 
> * [ ]  lectures/ar1_bayes.md:!pip install arviz pymc numpyro jax
> * [ ]  lectures/ar1_bayes.md:import jax.numpy as jnp
> * [ ]  lectures/ar1_bayes.md:from jax import random
> * [ ]  lectures/back_prop.md:!pip install --upgrade jax jaxlib
> * [ ]  lectures/back_prop.md:import jax.numpy as jnp
> * [ ]  lectures/back_prop.md:from jax import grad, jit, jacfwd, vmap
> * [ ]  lectures/back_prop.md:from jax import random
> * [ ]  lectures/back_prop.md:import jax
> * [ ]  lectures/back_prop.md:    h = jax.nn.sigmoid
> * [ ]  lectures/back_prop.md:D = jax.scipy.linalg.block_diag(*[row.reshape((1, 2)) for row in jnp.block([[δ * xs[:-1]], [δ]]).T])
> * [ ]  lectures/back_prop.md:dxs_la = jax.scipy.linalg.solve_triangular(jnp.eye(N) - L, D, lower=True)
> * [ ]  lectures/back_prop.md:    D = jax.scipy.linalg.block_diag(*[row.reshape((1, 2)) for row in jnp.block([[δ * xs[:-1]], [δ]]).T])
> * [ ]  lectures/back_prop.md:    dxs_la = jax.scipy.linalg.solve_triangular(jnp.eye(N) - L, D, lower=True)
> * [ ]  lectures/back_prop.md:from jax.lib import xla_bridge
> * [ ]  lectures/bayes_nonconj.md:!pip install numpyro pyro-ppl torch jax
> * [ ]  lectures/bayes_nonconj.md:# jax
> * [ ]  lectures/bayes_nonconj.md:import jax.numpy as jnp
> * [ ]  lectures/bayes_nonconj.md:from jax import lax, random
> * [ ]  lectures/bayes_nonconj.md:We use both the packages `pyro` and `numpyro` with assistance from  `jax` to approximate a  posterior distribution
> * [ ]  lectures/bayes_nonconj.md:        # jax requires explicit PRNG state to be passed
> * [x]  lectures/geom_series.md:init_printing(use_latex='mathjax')
> * [ ]  lectures/mix_model.md:import jax.numpy as jnp
> * [ ]  lectures/mix_model.md:from jax import random
> 
> We will need to do that before this can be merged. Can I get your help in coordinating this?

Hi @mmcky,

I will work on this now. I think only the intro series have been sanitized.

**Comment by HumphreyYang** [comment]
Hi @jstac and @mmcky,

I had a quick look at these lectures. Except `back_prop.md`, all other lectures use JAX just for `numpyro`:

`back_prop.md`:
I think back_prop is highly integrated with JAX. I can write a Pytorch/Tensorflow version of the lecture but it still needs GPU to run. Should we migrate this lecture as a whole to our JAX series?

`bayes_nonconj`:
This lecture can be divided into two -- one with the help of JAX `numpyro` and another without JAX `pyro`. Should we migrate JAX content to the JAX series for this lecture?

`mix_model.md`:
I can replicate the result of `numpyro` in  `pyro` without the use of JAX similar to what we can do for `bayes_nonconj` and migrate the JAX content to the JAX lecture.

`ar1_bayes`:
I can also replicate the result of `numpyro` in  `pyro` and migrate the JAX content to the JAX lecture.

Please kindly let me know your opinion on this. I think the lecture `back_prop.md` will take a very long time to run on CPU : )

**Comment by HumphreyYang** [comment]
> Roger that @HumphreyYang -- I had thought we were trying to move all `jax` code to `lecture-jax` but if that isn't correct let me know (cc: @jstac)
> 
> I don't think some of these lectures are in the `lecture-jax` series.

Many thanks @mmcky,

Yes, I think we mostly focus on the tough cases where we use JAX for the main algorithms, but we did not look into cases like these where JAX is used as a supporting tool.

**Comment by mmcky** [comment]
👍 thanks @HumphreyYang that is a really helpful summary of possible actions.

**Comment by mmcky** [comment]
Closing this as we will keep `jax` support in this lecture series.
